src/main/java/com/minecolonies/coremod/inventory/CraftingGUIBuilding.java

```diff
@@ -0,0 +1,350 @@
+                        return ItemStack.EMPTY;
+                    }
+
+                    @Override
+                    public boolean isItemValid(final ItemStack par1ItemStack)
+                    {
+                        return false;
+                    }
+
+                    @Override
+                    public boolean canTakeStack(final EntityPlayer par1EntityPlayer)
+                    {
+                        return false;
+                    }
+                });
+            }
+        }
+
+        int i;
+        for (i = 0; i < PLAYER_INVENTORY_ROWS; i++)
+        {
+            for (int j = 0; j < PLAYER_INVENTORY_COLUMNS; j++)
+            {
+                addSlotToContainer(new Slot(
+                        playerInventory,
+                        j + i * PLAYER_INVENTORY_COLUMNS + PLAYER_INVENTORY_COLUMNS,
+                        PLAYER_INVENTORY_INITIAL_X_OFFSET + j * INVENTORY_OFFSET_EACH,
+                        PLAYER_INVENTORY_INITIAL_Y_OFFSET + i * INVENTORY_OFFSET_EACH
+                ));
+            }
+        }
+
+        for (i = 0; i < PLAYER_INVENTORY_COLUMNS; i++)
+        {
+            addSlotToContainer(new Slot(
+                    playerInventory, i,
+                    PLAYER_INVENTORY_INITIAL_X_OFFSET + i * INVENTORY_OFFSET_EACH,
+                    PLAYER_INVENTORY_HOTBAR_OFFSET
+            ));
+        }
+
+        this.onCraftMatrixChanged(this.craftMatrix);
+    }
+
+    @Override
+    protected final Slot addSlotToContainer(final Slot slotToAdd)
+    {
+        return super.addSlotToContainer(slotToAdd);
+    }
+
+    /**
+     * Callback for when the crafting matrix is changed.
+     */
+    @Override
+    public void onCraftMatrixChanged(final IInventory inventoryIn)
+    {
+        if (!worldObj.isRemote)
+        {
+            EntityPlayerMP entityplayermp = (EntityPlayerMP) player;
+            ItemStack itemstack = ItemStack.EMPTY;
+            IRecipe irecipe = CraftingManager.findMatchingRecipe(craftMatrix, worldObj);
+            if (irecipe != null && (irecipe.isDynamic()
+                    || !worldObj.getGameRules().getBoolean("doLimitedCrafting")
+                    || entityplayermp.getRecipeBook().isUnlocked(irecipe)
+                    || entityplayermp.isCreative()))
+            {
+                this.craftResult.setRecipeUsed(irecipe);
+                itemstack = irecipe.getCraftingResult(this.craftMatrix);
+            }
+
+            this.craftResult.setInventorySlotContents(0, itemstack);
+            entityplayermp.connection.sendPacket(new SPacketSetSlot(this.windowId, 0, itemstack));
+        }
+
+        super.onCraftMatrixChanged(inventoryIn);
+    }
+
+    @Override
+    public boolean canInteractWith(final EntityPlayer playerIn)
+    {
+        return true;
+    }
+
+    @Override
+    public ItemStack slotClick(int slotId, int clickedButton, ClickType mode, EntityPlayer playerIn)
+    {
+        if (slotId >= 1 && slotId < 5)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 5 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)